docs/source/cuda/cuda_array_interface.rst

```diff
@@ -69,13 +69,213 @@ The following are optional entries:
+  :ref:`cuda-array-interface-synchronization` section below.
+
+- **sync**: ``bool``
+
+  Generally, if **sync** is given and is ``True``, then the consumer must either
+  synchronize on the provided **stream** at the point of consumption, or enqueue
+  operations on the data on the given **stream**. See
+  :ref:`cuda-array-interface-synchronization` for further details.
+
+
+.. _cuda-array-interface-synchronization:
+
+Synchronization
+---------------
+
+Definitions
+~~~~~~~~~~~
+
+When discussing synchronization, the following definitions are used:
+
+- *Producer*: The library / object on which ``__cuda_array_interface__`` is
+  accessed.
+- *Consumer*: The library / function that accesses the
+  ``__cuda_array_interface__`` of the Producer.
+- *User Code*: Code that induces a Producer and Consumer to share data through
+  the CAI.
+- *User*: The person writing or maintaining the User Code. The User may
+  implement User Code without knowledge of the CAI, since the CAI accesses can
+  be  hidden from their view.
+
+In the following example:
+
+.. code-block:: python
+
+   import cupy
+   from numba import cuda
+
+   @cuda.jit
+   def add(x, y, out):
+       start = cuda.grid(1)
+       stride = cuda.gridsize(1)
+       for i in range(start, x.shape[0], stride):
+           out[i] = x[i] + y[i]
+
+   a = cupy.arange(10)
+   b = a * 2
+   out = cupy.zeros_like(a)
+
+   add[1, 32](a, b, out)
+
+When the ``add`` kernel is launched:
+
+- ``a``, ``b``, ``c`` are Producers.
+- The ``add`` kernel is the Consumer.
+- The User Code is specifically ``add[1, 32](a, b, out)``.
+- The author of the code is the User.
+
+
+Design Motivations
+~~~~~~~~~~~~~~~~~~
+
+Elements of the CAI design related to synchronization seek to fulfil these
+requirements:
+
+1. Producers and Consumers that exchange data through the CAI must be able to do
+   so avoiding data races.
+2. Requirement 1 should be met without requiring the user to be
+   aware of any particulars of the CAI - in other words, exchanging data between
+   Producers and Consumers that operate on data asynchronously should be correct
+   by default.
+3. Where the User is aware of the particulars of the CAI and implementation
+   details of the Producer and Consumer, they should be able to, at their
+   discretion, override some of the synchronization semantics of the interface
+   to reduce the synchronization overhead. Overriding synchronization semantics
+   implies that:
+
+   - The CAI design, and the design and implementation of the Producer and
+     Consumer do not specify or guarantee correctness with respect to data
+     races.
+   - Instead, the User is responsible for ensuring correctness with respect to
+     data races.
+4. All operations and exchange of data are orchestrated from a single host
+   thread. Use of multiple threads on the host is outside of the scope of this
+   version of the CAI. A future revision to the interface specification may
```

Comment:
Do you see any specific issues arising from host multi-threading?  Or is this just to avoid scope creep?

Reply:
My guess is it's intended to leave the per-thread default stream out of scope?

Reply:
With @kkraus14's comment on the per-thread default stream (https://github.com/numba/numba/pull/5162#issuecomment-655535952), perhaps it is not an issue.

Reply:
I don't see any particular issues - I was just trying to keep the scope as small as possible for this revision.